proofs/tree.go

```diff
@@ -419,52 +435,58 @@ func CalculateProofNodeList(node, leafCount uint64) (nodes []*HashNode, err erro
 	height, _ := merkle.CalculateHeightAndNodeCount(leafCount)
 	index := 0
 	level := height - 1
-	lastNodeInLevel := leafCount-1
+	lastNodeInLevel := leafCount - 1
 	offset := uint64(0)
 	nodes = make([]*HashNode, height-1)
 
 	for i := level; i > 0; i-- {
 		// only add hash if this isn't an odd end
-		if !(node == lastNodeInLevel && (lastNodeInLevel+1) % 2 == 1) {
-			if node % 2 == 0 {
-				nodes[index] = &HashNode{false, offset+node+1}
+		if !(node == lastNodeInLevel && (lastNodeInLevel+1)%2 == 1) {
+			if node%2 == 0 {
+				nodes[index] = &HashNode{false, offset + node + 1}
 			} else {
-				nodes[index] = &HashNode{true, offset+node-1}
+				nodes[index] = &HashNode{true, offset + node - 1}
 			}
 			index++
 		}
-		node = node/2
+		node = node / 2
 
-		offset += lastNodeInLevel+1
-		lastNodeInLevel = (lastNodeInLevel+1) / 2 + (lastNodeInLevel+1) % 2 - 1
+		offset += lastNodeInLevel + 1
+		lastNodeInLevel = (lastNodeInLevel+1)/2 + (lastNodeInLevel+1)%2 - 1
 		level--
 	}
 	return nodes[:index], nil
 }
 
 // CalculateHashForProofField takes a Proof struct and returns a hash of the concatenated property name, value & salt.
 // Uses ConcatValues internally.
-func CalculateHashForProofField(proof *Proof) (hash []byte, err error) {
+func CalculateHashForProofField(proof *Proof, hashFunc hash.Hash) (hash []byte, err error) {
 	input, err := ConcatValues(proof.Property, proof.Value, proof.Salt)
 	if err != nil {
 		return []byte{}, err
 	}
 
-	h := blake2b.Sum256(input)
-	hash = h[:]
+	defer hashFunc.Reset()
+	_, err = hashFunc.Write(input[:])
+	if err != nil {
+		return []byte{}, err
+	}
+	hash = hashFunc.Sum(nil)
 	return hash, nil
 }
 
 // ValidateProofHashes calculates the merkle root based on a list of left/right hashes.
-func ValidateProofHashes(hash []byte, hashes []*MerkleHash, rootHash []byte) (valid bool, err error) {
+func ValidateProofHashes(hash []byte, hashes []*MerkleHash, rootHash []byte, hashFunc hash.Hash) (valid bool, err error) {
 	for i := 0; i < len(hashes); i++ {
 		if len(hashes[i].Left) == 0 {
-			hash = HashTwoValues(hash, hashes[i].Right)
+			hash = HashTwoValues(hash, hashes[i].Right, hashFunc)
 		} else {
-			hash = HashTwoValues(hashes[i].Left, hash)
+			hash = HashTwoValues(hashes[i].Left, hash, hashFunc)
 		}
 	}
 
+	fmt.Println("hash1", hash)
```

Comment:
Maybe we do not need this.

Reply:
Maybe? I think quite certainly we don't need it ;)